src/services/TestResultsService.ts

```diff
@@ -311,17 +311,45 @@ export class TestResultsService {
+                      if (registrationAnniversaryDifference >= 2 ) {
+                        payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                        // Checks that the Anniversary date is greater than 2 months from the test date - CVSB-11509 AC2
+                      } else if (registrationAnniversaryDifference < 2 && registrationAnniversaryDifference > 0) {
+                        payload = TestResultsService.generatesExpiryTwoYearsOnRegistrationDate(payload, index);
+                        // Checks that the Anniversary date is before the test date - CVSB-11509 AC3
+                      } else if (registrationAnniversaryDifference < 0) {
+                        payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                        // Fills the gap in logic if the registrationAnniversaryDifference == 0
+                      } else if (registrationAnniversaryDifference === 0) {
+                        // if the anniversaryDifference is between 0 to 0.4
+                        if (dateFns.isBefore(new Date(), registrationAnniversary)) {
+                          payload = TestResultsService.generatesExpiryTwoYearsOnRegistrationDate(payload, index);
+                          // if the anniversaryDifference is between 0 to -0.4
+                        } else {
+                          payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                        }
+                      }
+                  // Generates the expiry if there is no regnDate && the test isnt A COIF test type - CVSB-11509 AC4
+                  } else if (TestResultsService.isMostRecentExpiryNotFound(mostRecentExpiryDateOnAllTestTypesBySystemNumber) && !payload.regnDate && !COIF_EXPIRY_TEST_TYPES.IDS.includes(payload.testTypes[index].testTypeId)) {
+                    payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                  // Generates the expiry for COIF test types that have an expiry date - CVSB-11509 AC5
+                  } else if (!TestResultsService.isMostRecentExpiryNotFound(mostRecentExpiryDateOnAllTestTypesBySystemNumber) && COIF_EXPIRY_TEST_TYPES.IDS.includes(payload.testTypes[index].testTypeId)) {
+                    payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                  } else if ((TestResultsService.isMostRecentExpiryNotFound(mostRecentExpiryDateOnAllTestTypesBySystemNumber))
+                      || dateFns.isBefore(mostRecentExpiryDateOnAllTestTypesBySystemNumber, dateFns.startOfDay(new Date()))
+                      || dateFns.isAfter(mostRecentExpiryDateOnAllTestTypesBySystemNumber, dateFns.addMonths(new Date(), 2))) {
+                    payload = TestResultsService.generatesExpiryOneYearMinusDay(payload, index);
+                  } else if (dateFns.isToday(mostRecentExpiryDateOnAllTestTypesBySystemNumber)) {
```

Comment:
need to check valid date on this line

Reply:
As discussed, the issue is in the method `generatesExpiryOneYearOnRegistrationDate` which uses the regnDate in the payload.

Reply:
I rebased the branch with the latest changes on develop and added the check